### DISCUSSION TITLE
仓库的podman默认使用vfs文件系统，是不是需要改为fuse-overlayfs

### DISCUSSION THREAD

**Discussion by pmzgit** [linuxdeepin/developer-center#4621] [Developer Discussion | 开发者讨论]
v23 使用distrobox+podman 提供v20兼容方案，说实话比较推荐，至少我之前使用distrobox时，一直没找到deepin的镜像，现在docker hub上已经有 v20 和v23的镜像了，好评

但是v23 安装好podman时，默认使用vfs作为联合文件系统，其实podman rootless模式不推荐，应该使用 fuse-overlayfs

[podman/rootless_tutorial.md at main · containers/podman · GitHub](https://github.com/containers/podman/blob/main/docs/tutorials/rootless_tutorial.md)

所以不知道如下命令，官方如何搞定rootless模式

sudo apt install deepin-compatibility-mode

我已经讲vfs替换成 fuse-overlayfs，并用如下命令启动v20容器

distrobox create -i dockerproxy.com/linuxdeepin/apricot:v20.8-compatible -n deepin-v20 -v

不知道后期官方是不是优化dcm包

**Comment by BLumia** [comment]
cc @black-desk

**Comment by black-desk** [comment]
v23的kernel是比较新的 照理来说podman应该是会直接使用内核中的overlayfs才对 我不确定你说的vfs driver是啥 我之后会确认一下

**Reply by pmzgit** [reply]
> rootless 模式无法用内核 overlayfs，只能走 fuse，所以需要 fuse-overlayfs。可以考虑打包（如果还没有的话）+ Recommends

* 如肥猫大神所说，是用fuse-overlayfs，我贴出的链接是podman官方，对于rootless模式所需环境的配置。

* 同时，distrobox也明确说需要用podman的rootless模式

* 我内核用的是。6.3.3的内核，应该说比v23 默认的6.1内核新，不确定vfs？你podman info 看一下graphdirver(如果我没记错的话)

* v23包含fuse-overlayfs，之所以仓库的podman没使用 fuse-overlayfs，是因为storage.conf 里没有配置[storage] 相关配置（我发的链接里有相关说明），podman默认走了vfs，建议仓库打包storage.conf
* 我查看了目前仓库关于podman rootless模式需要的各依赖版本，都是支持的（cgroups v2启用，runc，slirp4netns， fuse-overlayfs等），所以你们的工作量并不大
* 最后我觉得你们把精力放在，用distrobox启动容器后，容器内的登录用户是宿主机上的uid，gid，而不是root（当然容器内的root实际上也是宿主机上的用户，也就是podman 启动镜像的用户）。因为我再v20上就是用的这个方案结合archlinux 容器日常使用的，容器内用户为root；不知道v23为什么映射的是宿主机登录用户（难道我第一次distrobox create容器时，加了 `--additional-flags "--pids-limit -1"` 选项，但是我把容器删了啊，搞不懂）但是目前可用
* **还有最重要的一点，distrobox我实测，官方的1.4.2.1，应该是有严重bug**，create的容器内，用户权限有问题，permission denied，不可用，我20和23都是用的main分支，才是正常的，希望仓库将来持续更新distrobox版本吧

**Reply by black-desk** [reply]
> 最后我觉得你们把精力放在，用distrobox启动容器后，容器内的登录用户是宿主机上的uid，gid，而不是root（当然容器内的root实际上也是宿主机上的用户，也就是podman 启动镜像的用户）。

这个是distrobox自己配的uid mapping, 和podman默认的行为模式不一样. 但是实际上更符合用户的使用习惯.

podman的人自己也有搞了个toolbox, 那玩意的uid mapping也是这么配的.

**Reply by black-desk** [reply]
> 还有最重要的一点，distrobox我实测，官方的1.4.2.1，应该是有严重bug，create的容器内，用户权限有问题，permission denied，不可用，我20和23都是用的main分支，才是正常的，希望仓库将来持续更新distrobox版本吧

你是指做什么操作提示权限不足了呢

**Reply by pmzgit** [reply]
任何操作都不可用，就相当于你不是当前用户

我说的不太准确，是由于非当前用户，才会permission denied；即，uid mapping的不对；
因为我已经装新版本了，没发复现，等我哪天再折腾个vm，复现一下，或者你可以用23仓库的distrobox，create一个容器试试，就知道了

**Reply by black-desk** [reply]
> 或者你可以用23仓库的distrobox，create一个容器试试，就知道了

我肯定有这么试过. 不确定你说的不是当前用户是什么意思.

**Reply by pmzgit** [reply]
> > 最后我觉得你们把精力放在，用distrobox启动容器后，容器内的登录用户是宿主机上的uid，gid，而不是root（当然容器内的root实际上也是宿主机上的用户，也就是podman 启动镜像的用户）。
> 
> 这个是distrobox自己配的uid mapping, 和podman默认的行为模式不一样. 但是实际上更符合用户的使用习惯.
> 
> podman的人自己也有搞了个toolbox, 那玩意的uid mapping也是这么配的.

我确认了一下，目前办公用v20.9也是普通用户、、、很奇怪，我记得去年测试20.9的时候，是root用户。不过记得当时distrobox貌似也不是1.4.2.1，应该是更早的版本。那就说明说的通了，可能后续版本映射改了。

不过1.4.2.1 creat的容器，uid 映射应该是有问题的，导致没有任何权限

**Comment by felixonmars** [comment]
rootless 模式无法用内核 overlayfs，只能走 fuse，所以需要 fuse-overlayfs。可以考虑打包（如果还没有的话）+ Recommends

**Reply by black-desk** [reply]
如果我没记错的话, rootless的overlayfs 在5.12还是5.15就已经可以用了. 我在arch上都用过(x

---

update: 在5.11就可用了

https://www.redhat.com/sysadmin/podman-rootless-overlay

**Comment by black-desk** [comment]
在v23的环境上看了一下。

podman 应该使用内核提供的rootless overlayfs, 但是实际上并没有用上。

storge driver 确实是vfs。

需要进一步看一下内核的配置和podman没有使用这个特性的具体原因。

---

debian sid 的行为和deepin v23一致。估计是什么内核的config没开。

refer: https://rootlesscontaine.rs/how-it-works/overlayfs/

@Zeno-sole

**Reply by Zeno-sole** [reply]
# Rootless OverlayFS 配置选项
#
# OverlayFS 是一种联合文件系统，可将两个或多个不同的文件系统以透明的方式叠加在一起。
# Rootless OverlayFS 是 OverlayFS 的一种变体，它允许普通用户在没有 root 权限的情况下使用 OverlayFS。
#
# 要启用 Rootless OverlayFS，请确保内核配置中包含以下选项。

CONFIG_OVERLAY_FS=y
CONFIG_OVERLAY_FS_REDIRECT_DIR=y
CONFIG_OVERLAY_FS_INDEX=y
CONFIG_OVERLAY_FS_XINO_AUTO=y
CONFIG_OVERLAY_FS_METACOPY=y
CONFIG_OVERLAY_FS_ALTHASH=y
CONFIG_OVERLAY_FS_ALLOW_REDIRECT=y
CONFIG_OVERLAY_FS_CONFIG_FILE=""
CONFIG_OVERLAY_FS_SNAPSHOT=y

# 这些选项使能够在没有 root 权限的情况下使用 OverlayFS。请注意，这些选项可能会有一些安全风险，
# 因此在生产环境中使用时应格外小心。

CONFIG_USER_NS=y
CONFIG_USERFAULTFD=y
CONFIG_NAMESPACES=y
CONFIG_PROC_PID_ARCH_STATUS=y
CONFIG_CHECKPOINT_RESTORE=y
CONFIG_USER_STACKTRACE_SUPPORT=y
CONFIG_BINFMT_MISC=y
CONFIG_UEVENT_HELPER_PATH="/sbin/hotplug"
CONFIG_BLK_DEV_DM=y
CONFIG_DM_THIN_PROVISIONING=y

# 在启用 Rootless OverlayFS 之前，还需要确保支持 user_namespaces。以下是需要启用的一些相关选项。

CONFIG_USER_NS=y
CONFIG_USER_NS_PROC=y
CONFIG_USERNS=y
CONFIG_USERFAULTFD=y
CONFIG_NAMESPACES=y
CONFIG_PID_NS=y
CONFIG_NET_NS=y

# 需要启用这些选项以支持 user_namespaces。这些选项也可能存在一些安全风险，请在使用时小心审查。

CONFIG_UTS_NS=y
CONFIG_IPC_NS=y
CONFIG_CGROUPS=y
CONFIG_CGROUP_CPUACCT=y
CONFIG_CGROUP_DEVICE=y
CONFIG_CGROUP_FREEZER=y
CONFIG_CGROUP_SCHED=y
CONFIG_CPUSETS=y
CONFIG_MEMCG=y
CONFIG_RESOURCE_COUNTERS=y
CONFIG_MEMCG_SWAP=y
CONFIG_MEMCG_KMEM=y
CONFIG_BLK_CGROUP=y
CONFIG_BLK_DEV_THROTTLING=y
CONFIG_CGROUP_PERF=y
CONFIG_CGROUP_HUGETLB=y
CONFIG_NETFILTER_NETLINK=y
CONFIG_NETFILTER_XTABLES=y
CONFIG_NETFILTER_XT_MATCH_ADDRTYPE=y
CONFIG_NETFILTER_XT_MATCH_CONNTRACK=y
CONFIG_NETFILTER_XT_MATCH_IPVS=y
CONFIG_NETFILTER_XT_MATCH_IPVS_IPV4=y
CONFIG_IP_NF_TARGET_MASQUERADE=y
CONFIG_IP6_NF_TARGET_MASQUERADE=y
CONFIG_NETFILTER_XT_MATCH_OWNER=y
CONFIG_NETFILTER_XT_MATCH_PKTTYPE=y
CONFIG_NETFILTER_XT_MATCH_QUOTA=y
CONFIG_NETFILTER_XT_MATCH_STATE=y
CONFIG_NETFILTER_XT_MATCH_TCPMSS=y
CONFIG_NETFILTER_XT_MATCH_U32=y
CONFIG_IP_NF_TARGET_REJECT=y
CONFIG_IP6_NF_TARGET_REJECT=y
CONFIG_NETFILTER_XT_TARGET_TCPMSS=y
CONFIG_NETFILTER_XT_TARGET_TEE=y
CONFIG_NETFILTER_XT_TARGET_TPROXY=y
CONFIG_NETFILTER_XT_TARGET_REDIRECT=y
CONFIG_NETFILTER_XT_TARGET
